pkg/lib/cockpit-upload-helper.js

```diff
@@ -0,0 +1,164 @@
+
+            if (this.resolveHandler) {
+                this.resolveHandler();
+                this.resolveHandler = null;
+            }
+        }
+    }
+
+    read_chunk(blob) {
+        return new Promise((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onload = (event) => resolve(event.target.result);
+            reader.onerror = reject;
+            reader.readAsArrayBuffer(blob);
+        });
+    }
+
+    write(chunk) {
+        console.assert(this.outstanding >= 0, "ack outstanding went negative", this.outstanding);
+        this.outstanding += 1;
+        this.channel.send(chunk);
+
+        if (this.outstanding < TOTAL_FRAME_WINDOW) {
+            return Promise.resolve();
+        } else {
+            return new Promise((resolve, reject) => {
+                this.resolveHandler = resolve;
+            });
+        }
+    }
+
+    close() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("close", function(event, message) {
+                if (message.problem) {
+                    reject(message);
+                } else {
+                    resolve();
+                }
+            });
+            channel.control({ command: "done" });
+        });
+    }
+
+    wait_ready() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("ready", function(event, message) {
+                resolve();
+            });
+        });
+    }
+
+    // Public methods
+
+    // Flush is waiting on acks in flight
+    async upload(file) {
+        let close_message = null;
+        this.channel.addEventListener("close", (event, message) => {
+            // If we are waiting on an "ack" and the disk is full
+            if (this.resolveHandler)
+                this.resolveHandler();
+            close_message = message;
+        });
+
+        await this.wait_ready();
+
+        let chunk_start = 0;
+        let send_chunks = 0;
+
+        while (this.channel.valid && chunk_start <= file.size) {
+            const chunk_next = chunk_start + BLOCK_SIZE;
+            const blob = file.slice(chunk_start, chunk_next);
+
+            const chunk = await this.read_chunk(blob);
+            await this.write(chunk);
+
+            send_chunks += blob.size;
+            if (this.progressCallback)
+                this.progressCallback(send_chunks);
+
+            chunk_start = chunk_next;
+        }
+
+        if (this.channel.valid) {
+            try {
+                await this.close();
+            } catch (exc) {
+                throw new Error(exc.problem);
+            }
+        } else if (close_message.problem) {
+            throw new Error(close_message.problem);
+        }
+    }
+
+    cancel() {
+        // Automatically handles reject for us
+        this.channel.close();
+    }
+}
```

Comment:
It might be nice to have a single function which takes the destination filename, file stream, progress callback and options blob and does all of the steps required and returns.

I think the object approach is still quite nice, of course: particularly for the `.cancel()` method.  I guess that could be an `AbortController` in function context... but that's definitely more than I'm asking for.

Reply:
Right, so a single function can't work because of the requirement of cancelling so it can' be a single function, unless you make it an object which in essence is a class.